codesyncapp/codesync-vs
# Fixed duplicate sending of diffs in case diffs are not processed within 5s

**Problem**
Daemon restarts every 5s and if diffs are not processed by server within 5s, plugin was sending same diffs again in every iteration. So if server took 20s to process the diffs, plugin actually sent same diffs 4 times

**Resolution**
We are now keeping track of diffs in a set called `diffsBeingProcessed` that are being processed by the Websocket client.
- This way same diffs are not sent over to server agaIn in each iteration
- Once a diff is successfully processed by server, we remove it from the `diffsBeingProcessed`
- If there is connection error raised from server, we empty this set so that we can retry upon new connection

**Testing steps**
- Run the server with single worker
- Add intentional blocking sleep in any endpoint of the server
- Call that endpoint from Postman
- Now try to send some diffs to server from Plugin
- Check logs of server that there are no duplicate diff files received from Plugin
- Verify from DB that there have been no recent duplicate diffs saved in DB, you can use following query
```
SELECT diff, COUNT(diff) c, created_at
FROM file_history
GROUP BY diff, created_at
HAVING c > 1
ORDER by created_at desc;
```
